File: ios/Hexgram/Models/Classics.swift
import Foundation

// MARK: - 数据类型

struct ClassicEntry {
    let name: String
    let judgment: String
    let yao: [String]
}

struct HuangjinceEntry {
    let category: String
    let label: String
    let text: String
}

struct QuestionCategory: Identifiable {
    let key: String
    let label: String
    var id: String { key }
}

struct ClassicsResult {
    let gaodao: ClassicEntry?
    let huangjince: HuangjinceEntry?
    let jiaoshi: String?
    let categories: [QuestionCategory]
}

// MARK: - 默认事类

let QUESTION_CATEGORIES: [QuestionCategory] = [
    QuestionCategory(key: "_总论", label: "综合"),
    QuestionCategory(key: "求财", label: "求财"),
    QuestionCategory(key: "事业", label: "事业"),
    QuestionCategory(key: "感情", label: "感情"),
    QuestionCategory(key: "婚姻", label: "婚姻"),
    QuestionCategory(key: "考试", label: "考试"),
    QuestionCategory(key: "家宅", label: "家宅"),
    QuestionCategory(key: "疾病", label: "疾病"),
    QuestionCategory(key: "出行", label: "出行"),
    QuestionCategory(key: "诉讼", label: "诉讼"),
    QuestionCategory(key: "失物", label: "失物"),
    QuestionCategory(key: "天气", label: "天气"),
    QuestionCategory(key: "怀孕", label: "怀孕"),
    QuestionCategory(key: "投资", label: "投资"),
    QuestionCategory(key: "求职", label: "求职"),
    QuestionCategory(key: "生意", label: "生意"),
]

// MARK: - 经典文献网络服务

/// 通过 Worker /api/classics 查询经典文献
/// Worker 先查 D1 缓存，缓存未命中则调用 LLM 生成并存入 D1
class ClassicsService {
    static let shared = ClassicsService()

    func query(
        guaKey: String,
        guaName: String = "",
        changedGuaKey: String? = nil,
        changedGuaName: String? = nil,
        category: String? = nil
    ) async throws -> ClassicsResult {
        let config = AIConfig.load()
        guard !config.endpoint.isEmpty else {
            return ClassicsResult(gaodao: nil, huangjince: nil, jiaoshi: nil, categories: QUESTION_CATEGORIES)
        }

        let baseUrl = config.endpoint.hasSuffix("/")
            ? String(config.endpoint.dropLast())
            : config.endpoint
        let classicsUrl = "\(baseUrl)/api/classics"

        guard let url = URL(string: classicsUrl) else {
            return ClassicsResult(gaodao: nil, huangjince: nil, jiaoshi: nil, categories: QUESTION_CATEGORIES)
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 60

        var body: [String: Any] = ["guaKey": guaKey, "guaName": guaName]
        if let ck = changedGuaKey { body["changedGuaKey"] = ck }
        if let cn = changedGuaName { body["changedGuaName"] = cn }
        if let cat = category { body["category"] = cat }
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            return ClassicsResult(gaodao: nil, huangjince: nil, jiaoshi: nil, categories: QUESTION_CATEGORIES)
        }

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return ClassicsResult(gaodao: nil, huangjince: nil, jiaoshi: nil, categories: QUESTION_CATEGORIES)
        }

        return parseResponse(json)
    }

    private func parseResponse(_ json: [String: Any]) -> ClassicsResult {
        // gaodao
        var gaodao: ClassicEntry? = nil
        if let g = json["gaodao"] as? [String: Any] {
            let yao = (g["yao"] as? [String]) ?? []
            gaodao = ClassicEntry(
                name: g["name"] as? String ?? "",
                judgment: g["judgment"] as? String ?? "",
                yao: yao
            )
        }

        // huangjince
        var huangjince: HuangjinceEntry? = nil
        if let h = json["huangjince"] as? [String: Any] {
            huangjince = HuangjinceEntry(
                category: h["category"] as? String ?? "",
                label: h["label"] as? String ?? "",
                text: h["text"] as? String ?? ""
            )
        }

        // jiaoshi
        let jiaoshi = json["jiaoshi"] as? String

        // categories
        var cats = QUESTION_CATEGORIES
        if let catArr = json["categories"] as? [[String: Any]], !catArr.isEmpty {
            cats = catArr.map {
                QuestionCategory(
                    key: $0["key"] as? String ?? "",
                    label: $0["label"] as? String ?? ""
                )
            }
        }

        return ClassicsResult(gaodao: gaodao, huangjince: huangjince, jiaoshi: jiaoshi, categories: cats)
    }
}

File: android/src/main/kotlin/com/hexgram/android/models/Classics.kt
package com.hexgram.android.models

import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

// 数据类
data class ClassicEntry(
    val name: String,
    val judgment: String,
    val yao: List<String>
)

data class HuangjinceEntry(
    val category: String,
    val label: String,
    val text: String
)

data class QuestionCategory(
    val key: String,
    val label: String
)

data class ClassicsResult(
    val gaodao: ClassicEntry?,
    val huangjince: HuangjinceEntry?,
    val jiaoshi: String?,
    val categories: List<QuestionCategory>
)

// 默认事类（在网络不可用时使用）
val DEFAULT_CATEGORIES = listOf(
    QuestionCategory("_总论", "综合"),
    QuestionCategory("求财", "求财"),
    QuestionCategory("事业", "事业"),
    QuestionCategory("感情", "感情"),
    QuestionCategory("婚姻", "婚姻"),
    QuestionCategory("考试", "考试"),
    QuestionCategory("家宅", "家宅"),
    QuestionCategory("疾病", "疾病"),
    QuestionCategory("出行", "出行"),
    QuestionCategory("诉讼", "诉讼"),
    QuestionCategory("失物", "失物"),
    QuestionCategory("天气", "天气"),
    QuestionCategory("怀孕", "怀孕"),
    QuestionCategory("投资", "投资"),
    QuestionCategory("求职", "求职"),
    QuestionCategory("生意", "生意"),
)

/**
 * 经典文献服务 - 通过 Worker /api/classics 查询
 * Worker 先查 D1 缓存，缓存未命中则调用 LLM 生成并存入 D1
 */
object ClassicsService {

    suspend fun query(
        context: Context,
        guaKey: String,
        guaName: String = "",
        changedGuaKey: String? = null,
        changedGuaName: String? = null,
        category: String? = null
    ): ClassicsResult = withContext(Dispatchers.IO) {
        val endpoint = AIService.getEndpoint(context)
        if (endpoint.isBlank()) {
            return@withContext ClassicsResult(null, null, null, DEFAULT_CATEGORIES)
        }

        // 构建 /api/classics URL
        val baseUrl = endpoint.trimEnd('/')
        val classicsUrl = if (baseUrl.endsWith("/api/classics")) baseUrl
            else baseUrl.replace(Regex("/+$"), "").let {
                // 如果 endpoint 是 "https://xxx.workers.dev" 或 "https://xxx.workers.dev/"
                val base = it.removeSuffix("/")
                "$base/api/classics"
            }

        val conn = URL(classicsUrl).openConnection() as HttpURLConnection
        try {
            conn.requestMethod = "POST"
            conn.setRequestProperty("Content-Type", "application/json")
            conn.connectTimeout = 15_000
            conn.readTimeout = 60_000
            conn.doOutput = true

            val body = JSONObject().apply {
                put("guaKey", guaKey)
                put("guaName", guaName)
                if (changedGuaKey != null) put("changedGuaKey", changedGuaKey)
                if (changedGuaName != null) put("changedGuaName", changedGuaName)
                if (category != null) put("category", category)
            }

            OutputStreamWriter(conn.outputStream, "UTF-8").use { it.write(body.toString()) }

            val code = conn.responseCode
            val text = if (code in 200..299) {
                conn.inputStream.bufferedReader().use { it.readText() }
            } else {
                return@withContext ClassicsResult(null, null, null, DEFAULT_CATEGORIES)
            }

            parseClassicsResponse(JSONObject(text))
        } catch (e: Exception) {
            ClassicsResult(null, null, null, DEFAULT_CATEGORIES)
        } finally {
            conn.disconnect()
        }
    }

    private fun parseClassicsResponse(json: JSONObject): ClassicsResult {
        // gaodao
        val gaodao = json.optJSONObject("gaodao")?.let { g ->
            val yaoArr = g.optJSONArray("yao")
            val yao = if (yaoArr != null) {
                (0 until yaoArr.length()).map { yaoArr.optString(it, "") }
            } else emptyList()
            ClassicEntry(
                name = g.optString("name", ""),
                judgment = g.optString("judgment", ""),
                yao = yao
            )
        }

        // huangjince
        val huangjince = json.optJSONObject("huangjince")?.let { h ->
            HuangjinceEntry(
                category = h.optString("category", ""),
                label = h.optString("label", ""),
                text = h.optString("text", "")
            )
        }

        // jiaoshi
        val jiaoshi = json.optString("jiaoshi", "").ifBlank { null }

        // categories
        val catArr = json.optJSONArray("categories")
        val categories = if (catArr != null && catArr.length() > 0) {
            (0 until catArr.length()).map { i ->
                val c = catArr.getJSONObject(i)
                QuestionCategory(c.optString("key"), c.optString("label"))
            }
        } else DEFAULT_CATEGORIES

        return ClassicsResult(gaodao, huangjince, jiaoshi, categories)
    }
}
